### DISCUSSION TITLE
How to add custom materials?

### DISCUSSION THREAD

**Discussion by Lancelotbronner** [SigmundGranaas/forgero#398] [General]
I'm making a modpack and trying to add new materials. I'm using KubeJS's data folder to register a new forgero pack with the materials but it never gets registered.

I've added the following file in `data/forgero/packs/halstein/materials/flint.json`

```json
{
	"type": "MINERAL",
	"name": "flint",
	"container": {
		"type": "HOST",
		"id": "minecraft:flint"
	},
	"properties": {
		"attributes": []
	},
	"palette": {
		"name": "stone"
	}
}
```

I've also added `package.json` and `default.json` in `data/forgero/packs/halstein`

```json
{
	"name": "halstein",
	"resource_type": "PACKAGE",
	"priority": 8,
	"namespace": "halstein",
	"dependencies": [
	  "minecraft",
	  "forgero",
	  "minecraft-vanilla"
	]
  }
```

```json
{
	"namespace": "halstein",
	"dependencies": [
		"minecraft",
		"forgero"
	]
}
```

What am I missing?

**Comment by Lancelotbronner** [comment]
I have also tried what is currently in the wiki, placing the material file in `data/forgero/materials/flint.json` but that doesn't seem to work either.

**Comment by SigmundGranaas** [comment]
Hi! And thanks for trying out the new data system! The docs on the wiki are outdated and should be removed not to confuse anyone. You did everything correctly it seems, but the only thing that is missing is the fact the the datapack has to be packaged as a mod, that is, inside a jar. The reason for this is that vanilla datapacks are loaded after the item registry is frozen. This means that datapacks cannot add resources that would result in new items being created. Therefore datapacks can only update the stats of currently registered resources.

It might be possible to look at if kubejs could be used as a lightweight entrypoint for adding new resources, but I haven't looked at it enough to know for sure.

Are you familiar with how to create basic mods, or would you like me to create an example for how it could be done?

**Reply by Lancelotbronner** [reply]
I have made attempts and could probably figure it out, I wonder if it would be possible to package a mod without an entrypoint, just to house the data. Ooooh I'll have to check with Quilt or Fabric!

**Reply by Lancelotbronner** [reply]
Alright, so I believe I have gotten the mod side of things working as it shows up without crashing the game. I've just moved that data to the resources folder, so I have `src/main/resources/data/forgero/packs/halstein` as the root folder for that pack above. That side of things doesn't seem to be registered.

| Index | Mod                                               | ID                                            | Version                      | Plugin                 | File(s)                                                                                                                                         | Sub-Files                                                                            | Sub-Files                                                        | Sub-Files                                                   |
|------:|---------------------------------------------------|-----------------------------------------------|------------------------------|------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------|------------------------------------------------------------------|-------------------------------------------------------------|
|   137 | Halstein                                          | halstein                                      | 1.0.0+1.19.2                 | quilt_loader           | <mods>\halstein-1.0.0+1.19.2.jar                                                                                                                |                                                                                      |                                                                  |                                                             |

On a side-note, why is `forgero-vanilla` there twice? Here's the log lines for enabled packs:

```log
[15:22:09] [Render thread/INFO]: Registered and validated 15 Forgero packages
[15:22:09] [Render thread/INFO]: [hoe-additions, sword-additions, pickaxe-additions, minecraft-extended, forgero-trinkets, forgero-extended, ancient-relics, shovel-additions, axe-additions, minecraft-wood-variants, minecraft-stone-variants, minecraft-variants, minecraft-vanilla, forgero-vanilla, forgero-vanilla]
```

I'm also getting these spammed but I'm pretty sure they're harmless, albeit a bit annoying

```log
[15:22:27] [Worker-Main-6/WARN]: Multiple paths added for '/assets/forgero/templates/textures/square_shovel_head_tip_reinforcement.png', but only a single one can be returned!
 - QuiltZipFileSystem[transformed-mod-forgero-extended.i0] /assets/forgero/templates/textures/square_shovel_head_tip_reinforcement.png
 - QuiltZipFileSystem[transformed-mod-forgero-compat.i0] /assets/forgero/templates/textures/square_shovel_head_tip_reinforcement.png
```

**Reply by SigmundGranaas** [reply]
Yeah it should not need an entrypoint as long as the mod has resources in the right folder. I don't know why forgero-vanilla gets registered twice, there might be some resource duplication going on. I also haven't seems the quilt error before, as I am only using fabric but I should look into it.

Did you get the resource registered in the system? Oops, just saw that it did not register.

**Reply by Lancelotbronner** [reply]
The resources did not seem to register no, I'll check if they're actually in the built jar

**Reply by Lancelotbronner** [reply]
The jar contains the resources at `/data/forgero/packs/halstein` but they aren't registered to forgero

**Reply by SigmundGranaas** [reply]
Resource location and pack placement looks good, so that should not be an issue. I haven't tried adding resources from quilt mods yet, so that is still uncharted territory.

**Reply by SigmundGranaas** [reply]
I will set up a demo repository with a example mod just to validate that I have a setup that works. I'll also try out a basic quilt setup as it's possible that's where the issue is coming from. I don't know when I'll have that ready, but I'll keep you updated. Hopefully in a couple of days

**Reply by Lancelotbronner** [reply]
Sounds good!

**Reply by SigmundGranaas** [reply]
I think the missing piece was adding the annotation to signal this is supposed to be used as a forgero resource. If using fabric, the fabric.mod.json file needs to contains this:
`  "custom": {
    "forgeroResource": true
    }`

**Reply by Lancelotbronner** [reply]
Thank you! I'll try this today, first with quilt.mod.json and if that doesn't work I'll make a fabric mod!

**Reply by SigmundGranaas** [reply]
It seems like the current version has an issue with picking up the material palette in the example. This results in all parts from a material being invisible. I don't know why this happens, but I am looking into it.

**Comment by SigmundGranaas** [comment]
Here's a very simple example repo for adding custom resources: https://github.com/SigmundGranaas/forgero-example-datapack

**Reply by SigmundGranaas** [reply]
I'll look into a quilt one as well, but it might take a little bit of time before I have the time to do that. The example repo is the absolute bare minimum for now, so if you have some suggestions for which example resources to add, feel free to tell me!

**Reply by SigmundGranaas** [reply]
I tried settign up a quilt example pack, but I can't get it to pick up the resources. I believe it is because I am using the fabric loader to load potential forgero pack candidates, but I am not sure. I'll leave that one be for the time being, as I fixed all of the issues with the fabric version.

**Reply by SigmundGranaas** [reply]
Ooopsie, I take that back, Quilt works now as well. The customdata needed in the mod.json file requirement has been removed, but I will probably bring it back if it leads to performance regressions.

**Reply by SigmundGranaas** [reply]
And here's the quilt example: https://github.com/SigmundGranaas/forgero-example-datapack-quilt
As expected it only works using the quilt loader.

**Reply by Lancelotbronner** [reply]
Switching loaders was my next test, thanks! I'll try the quilt solution and let you know how things go!

**Reply by Lancelotbronner** [reply]
I had tested a fabric mod on quilt but it really is specific to the loader rather than the API

**Reply by SigmundGranaas** [reply]
Also, we have started a channel on discord for creating and configuring datapacks for Forgero. If you're interested in creating extension packs, you should definitely join the discussion over there!

**Reply by Lancelotbronner** [reply]
I will!

**Comment by Lancelotbronner** [comment]
Custom materials have been added, thanks a lot! I've been trying to use the autogenerating palette feature via

```json
"palette": {
	"name": "flint",
	"include": [
		"minecraft:textures/item/flint.png"
	],
	"exclude": []
}
```

but it doesn't appear to work, I've also tried removing `.png` from the resource location. Is this still supported?

**Reply by SigmundGranaas** [reply]
Unfortunately, palette generators were deprecated in earlier versions because creating the palette manually was more consistent, and actually quite easy. Re-importing the generated palettes, was such a hassle I preferred crating them manually. I can recreate the functionality though if it would be convenient for people making extension packs.

**Reply by Lancelotbronner** [reply]
Could you point me to the code? I could translate it into a CLI tool in a language I'm more comfortable with

**Reply by SigmundGranaas** [reply]
https://github.com/SigmundGranaas/forgero/blob/1.18.2-0.9.1/src/main/java/com/sigmundgranaas/forgero/core/texture/palette/PaletteFactoryImpl.java


The tricky part is being able to provide the factory with the correct buffered images, it used Minecraft's own asset loader to fetch the relevant images when stitching textures.
